drivers/ethernet/eth_xmc4xxx.c

```diff
@@ -0,0 +1,1144 @@
+
+	for (int i = 0; i < NUM_RX_DMA_DESCRIPTORS; i++) {
+		XMC_ETH_MAC_DMA_DESC_t *dma_desc = &rx_dma_desc[i];
+		struct net_buf *rx_buf = net_pkt_get_reserve_rx_data(CONFIG_NET_BUF_DATA_SIZE,
+								     K_NO_WAIT);
+
+		if (rx_buf == NULL) {
+			free_rx_bufs(dev);
+			LOG_ERR("Failed to reserve data net buffers");
+			return -ENOBUFS;
+		}
+
+		dev_data->rx_frag_list[i] = rx_buf;
+		dma_desc->buffer1 = (uint32_t)rx_buf->data;
+		dma_desc->length = rx_buf->size | ETH_RX_DMA_DESC_SECOND_ADDR_CHAINED_MASK;
+		dma_desc->status = ETH_MAC_DMA_RDES0_OWN;
+	}
+
+	return 0;
+}
+
+static int eth_xmc4xxx_reset(const struct device *dev)
+{
+	const struct eth_xmc4xxx_config *dev_cfg = dev->config;
+	int retries = ETH_RESET_RETRIES;
+	int sleep_per_retry = ETH_RESET_TIMEOUT_USEC / ETH_RESET_RETRIES;
+
+	dev_cfg->regs->BUS_MODE |= ETH_BUS_MODE_SWR_Msk;
+
+	/* reset may fail if the clocks are not properly setup */
+	for (int i = 0; i < retries; i++) {
+		if ((dev_cfg->regs->BUS_MODE & ETH_BUS_MODE_SWR_Msk) == 0) {
+			return 0;
+		}
+		k_busy_wait(sleep_per_retry);
+	}
+
+	return -EIO;
+}
+
+static inline void eth_xmc4xxx_set_mac_address(ETH_GLOBAL_TypeDef *regs, uint8_t *const addr)
+{
+	regs->MAC_ADDRESS0_HIGH = addr[4] | (addr[5] << 8);
+	regs->MAC_ADDRESS0_LOW = addr[0] | (addr[1] << 8) | (addr[2] << 16) | (addr[3] << 24);
+}
+
+static inline void eth_xmc4xxx_mask_unused_interrupts(ETH_GLOBAL_TypeDef *regs)
+{
+	/* Disable Mac Management Counter (MMC) interrupt events */
+	regs->MMC_TRANSMIT_INTERRUPT_MASK = ETH_MAC_DISABLE_MMC_INTERRUPT_MSK;
+	regs->MMC_RECEIVE_INTERRUPT_MASK = ETH_MAC_DISABLE_MMC_INTERRUPT_MSK;
+
+	/* IPC - Receive IP checksum checker */
+	regs->MMC_IPC_RECEIVE_INTERRUPT_MASK = ETH_MAC_DISABLE_MMC_IPC_RECEIVE_INTERRUPT_MSK;
+
+	/* Disable PMT and timestamp interrupt events */
+	regs->INTERRUPT_MASK = ETH_INTERRUPT_MASK_PMTIM_Msk | ETH_INTERRUPT_MASK_TSIM_Msk;
+}
+
+static inline void eth_xmc4xxx_init_timestamp_control_reg(ETH_GLOBAL_TypeDef *regs)
+{
+#if defined(CONFIG_NET_GPTP)
+	regs->TIMESTAMP_CONTROL = ETH_TIMESTAMP_CONTROL_TSENA_Msk |
+				  ETH_TIMESTAMP_CONTROL_TSENALL_Msk;
+#endif
+
+#if defined(CONFIG_PTP_CLOCK_XMC4XXX)
+	/* use fine control */
+	regs->TIMESTAMP_CONTROL |= ETH_TIMESTAMP_CONTROL_TSCFUPDT_Msk |
+				  ETH_TIMESTAMP_CONTROL_TSCTRLSSR_Msk;
+
+	/* make ptp run at 50MHz - implies 20ns increment for each increment of the */
+	/* sub_second_register */
+	regs->SUB_SECOND_INCREMENT = 20;
+
+	/* f_out = f_cpu * K / 2^32, where K = TIMESTAMP_ADDEND. Target F_out = 50MHz  */
+	/* Therefore, K = ceil(f_out * 2^32 / f_cpu) */
+
+	uint32_t f_cpu = XMC_SCU_CLOCK_GetSystemClockFrequency();
+	uint32_t K = (BIT64(32) * ETH_PTP_CLOCK_FREQUENCY  + f_cpu / 2) / f_cpu;
+
+	regs->TIMESTAMP_ADDEND = K;
+
+	/* Addend register update */
+	regs->TIMESTAMP_CONTROL |= ETH_TIMESTAMP_CONTROL_TSADDREG_Msk;
+	while (regs->TIMESTAMP_CONTROL & ETH_TIMESTAMP_CONTROL_TSADDREG_Msk) {
```

Comment:
Could we use `WAIT_FOR` macro here and few lines below?

Reply:
thanks, will update.